File: MarkdownIOS/SimpleMarkdownParser/Classes/Core/SimpleMarkdownParser.swift
//
//  SimpleMarkdownParser.swift
//  SimpleMarkdownParser Pod
//
//  Core library: defines the protocol of the parser
//

// The enum to define the way to extract the text between 2 tags
public enum ExtractBetweenMode {
    
    case startToNext
    case intermediateToNext
    case intermediateToEnd
    
}

// Parser protocol
public protocol SimpleMarkdownParser: class {
    
    func findTags(_ markdownText: String) -> [MarkdownTag]
    func extractText(_ markdownText: String, tag: MarkdownTag) -> String
    func extractTextBetween(_ markdownText: String, startTag: MarkdownTag, endTag: MarkdownTag, mode: ExtractBetweenMode) -> String
    func extractFull(_ markdownText: String, tag: MarkdownTag) -> String
    func extractFullBetween(_ markdownText: String, startTag: MarkdownTag, endTag: MarkdownTag, mode: ExtractBetweenMode) -> String
    func extractExtra(_ markdownText: String, tag: MarkdownTag) -> String
    
}

File: MarkdownAndroid/SimpleMarkdownParserLib/src/main/java/com/crescentflare/simplemarkdownparser/core/SimpleMarkdownParser.java
package com.crescentflare.simplemarkdownparser.core;

/**
 * Simple markdown parser core library: interface class
 * The interface to do the core (low-level) markdown parsing
 * It returns ranges for the markdown tags which is used within the library
 * Use manually if the output needs to be highly customizable
 */
public interface SimpleMarkdownParser
{
    enum ExtractBetweenMode
    {
        StartToNext,
        IntermediateToNext,
        IntermediateToEnd
    }

    MarkdownTag[] findTags(String markdownText);
    String extractText(String markdownText, MarkdownTag tag);
    String extractTextBetween(String markdownText, MarkdownTag startTag, MarkdownTag endTag, ExtractBetweenMode mode);
    String extractFull(String markdownText, MarkdownTag tag);
    String extractFullBetween(String markdownText, MarkdownTag startTag, MarkdownTag endTag, ExtractBetweenMode mode);
    String extractExtra(String markdownText, MarkdownTag tag);
}
